trie/smt.py

```diff
@@ -0,0 +1,355 @@
+            # sibling node in the source branch at the branch point.
+            # NOTE: Keys are in MSB->LSB (root->leaf) order.
+            #       Node lists are in root->leaf order.
+            #       Be sure to convert between them effectively.
+            for bit in reversed(range(self._branch_size)):
+                if path_diff & (1 << bit) > 0:
+                    branch_point = (self._branch_size-1)-bit
+                    break
+
+            # NOTE: node_updates only has to be as long as necessary
+            #       to obtain the update. This allows an optimization
+            #       of pruning updates to the maximum possible depth
+            #       that would be required to merge, which may be many
+            #       orders of magnitude smaller than the tree depth.
+            if len(node_updates) <= branch_point:
+                raise ValidationError("Updated node list is not deep enough")
+
+            # Update sibling in the branch where our key differs from the update
+            self._branch[branch_point] = node_updates[branch_point]
+            # No need to update value
+
+
+class SparseMerkleTree:
+
+    def __init__(self, key_size: int=32, default: bytes=BLANK_NODE):
+        # Ensure we can support the given depth
+        if not 1 <= key_size <= 32:
+            raise ValidationError("Keysize must be number of bytes in range [0, 32]")
+
+        self._key_size = key_size  # key size in bytes
+        self.depth = key_size * 8  # depth is number of bits in key
+
+        self._default = default
+
+        # Initialize an empty tree with one branch
+        self.db = {}
+        node = self._default  # Default leaf node
+        for _ in range(self.depth):
+            node_hash = keccak(node)
+            self.db[node_hash] = node
+            node = node_hash + node_hash
+
+        # Finally, write the root hash
+        self.root_hash = keccak(node)
+        self.db[self.root_hash] = node
+
+    @classmethod
+    def from_db(
+            cls,
+            db: Dict[bytes, bytes],
+            root_hash: Hash32,
+            key_size: int=32,
+            default: bytes=BLANK_NODE):
+        smt = cls(key_size=key_size, default=default)
+
+        # If db is provided, and is not consistent,
+        # there may be a silent error. Can't solve that easily.
+        for key, value in db.items():
+            validate_is_bytes(key)
+            validate_length(key, 32)  # Must be a bytes32 hash
+            validate_is_bytes(value)
+            smt.db[key] = value
+
+        # Set root_hash, so we know where to start
+        validate_is_bytes(root_hash)
+        validate_length(root_hash, 32)  # Must be a bytes32 hash
+
+        smt.root_hash = root_hash
+
+        return smt
+
+    def get(self, key: bytes) -> bytes:
+        value, _ = self._get(key)
+
+        # Ensure that it isn't blank!
+        if value == BLANK_NODE:
+            raise KeyError("Key does not exist")
+
+        return value
```

Comment:
Does that lead to an optimization or better code analysis?

Reply:
It's all about readability: the visual confirmation that this is a branching situation and the lines below it only happen in the alternate case.

Reply:
@carver Maybe too many cooks here, but isn't it also common for guarding (i.e. stack popping) if statements to lack else clauses?  I know that I often write code in this pattern and sort of prefer it since it's not hard to tell at a glance that the body of the if statement would prevent any further execution in that subroutine.